File: packages/pvm-assemblyscript/assembly/instructions/arithmetic-64.ts
import { OPCODE_ADD_64, OPCODE_DIV_S_64, OPCODE_DIV_U_64, OPCODE_MUL_64, OPCODE_REM_S_64, OPCODE_REM_U_64, OPCODE_SUB_64 } from '../config'
import { InstructionContext, InstructionResult } from '../types'
import { BaseInstruction } from './base'

export class ADD_64Instruction extends BaseInstruction {
  opcode: i32 = OPCODE_ADD_64
  name: string = 'ADD_64'
  
  execute(context: InstructionContext): InstructionResult {
    const registerD = this.getRegisterD(context.operands)
    const registerA = this.getRegisterA(context.operands)
    const registerB = this.getRegisterB(context.operands)
    const valueA = this.getRegisterValueAs64(context.registers, registerA)
    const valueB = this.getRegisterValueAs64(context.registers, registerB)
    const result = valueA + valueB

    this.setRegisterValueWith64BitResult(context.registers, registerD, result)

    return new InstructionResult(-1)
  }
}

export class SUB_64Instruction extends BaseInstruction {
  opcode: i32 = OPCODE_SUB_64
  name: string = 'SUB_64'
  
  execute(context: InstructionContext): InstructionResult {
    const registerD = this.getRegisterD(context.operands)
    const registerA = this.getRegisterA(context.operands)
    const registerB = this.getRegisterB(context.operands)
    const valueA = this.getRegisterValueAs64(context.registers, registerA)
    const valueB = this.getRegisterValueAs64(context.registers, registerB)
    const result = valueA - valueB

    this.setRegisterValueWith64BitResult(context.registers, registerD, result)

    return new InstructionResult(-1)
  }
}

export class MUL_64Instruction extends BaseInstruction {
  opcode: i32 = OPCODE_MUL_64
  name: string = 'MUL_64'
  
  execute(context: InstructionContext): InstructionResult {
    const registerD = this.getRegisterD(context.operands)
    const registerA = this.getRegisterA(context.operands)
    const registerB = this.getRegisterB(context.operands)
    const valueA = this.getRegisterValueAs64(context.registers, registerA)
    const valueB = this.getRegisterValueAs64(context.registers, registerB)
    const result = valueA * valueB

    this.setRegisterValueWith64BitResult(context.registers, registerD, result)

    return new InstructionResult(-1)
  }
}

export class DIV_U_64Instruction extends BaseInstruction {
  opcode: i32 = OPCODE_DIV_U_64
  name: string = 'DIV_U_64'
  
  execute(context: InstructionContext): InstructionResult {
    const registerD = this.getRegisterD(context.operands)
    const registerA = this.getRegisterA(context.operands)
    const registerB = this.getRegisterB(context.operands)
    const valueA = this.getRegisterValueAs64(context.registers, registerA)
    const valueB = this.getRegisterValueAs64(context.registers, registerB)

    // Gray Paper: division by zero results in 2^64 - 1
    // Use u64 to avoid overflow: 0xFFFFFFFFFFFFFFFF
    const result = valueB === i64(0) ? i64(u64(0xFFFFFFFFFFFFFFFF)) : valueA / valueB

    this.setRegisterValue(context.registers, registerD, result)

    return new InstructionResult(-1)
  }
}

export class DIV_S_64Instruction extends BaseInstruction {
  opcode: i32 = OPCODE_DIV_S_64
  name: string = 'DIV_S_64'
  
  execute(context: InstructionContext): InstructionResult {
    const registerD = this.getRegisterD(context.operands)
    const registerA = this.getRegisterA(context.operands)
    const registerB = this.getRegisterB(context.operands)
    const valueA = this.getRegisterValueAs64(context.registers, registerA)
    const valueB = this.getRegisterValueAs64(context.registers, registerB)

    // Convert to signed values
    const signedA = this.toSigned64(valueA)
    const signedB = this.toSigned64(valueB)

    // Gray Paper: handle special cases
    let result: i64
    if (signedB === i64(0)) {
      // Division by zero: result = 2^64 - 1
      // Use u64 to avoid overflow: 0xFFFFFFFFFFFFFFFF
      result = i64(u64(0xFFFFFFFFFFFFFFFF))
    } else if (signedA === -(i64(1) << i64(63)) && signedB === i64(-1)) {
      // Overflow case: result = unsigned(a) = valueA
      result = valueA
    } else {
      // Normal division: round towards zero
      // Gray Paper: unsigned{rtz(a ÷ b)} - convert signed result to unsigned
      const signedResult = signedA / signedB
      // Convert negative result to unsigned: add 2^64
      // Use u64 arithmetic to avoid overflow: signedResult + 2^64 = signedResult + 0xFFFFFFFFFFFFFFFF + 1
      if (signedResult < i64(0)) {
        // For negative values, we need to add 2^64 to get the unsigned representation
        // Since we can't do i64 + 2^64 directly, we use u64 arithmetic
        // u64(signedResult) when signedResult is negative gives a large positive (two's complement)
        // But we want signedResult + 2^64, which is: u64(signedResult) (already includes the 2^64 offset)
        // Actually, in two's complement, a negative i64 value when viewed as u64 IS the unsigned representation
        // So we can just cast: u64(signedResult) is correct
        result = i64(u64(signedResult))
      } else {
        result = signedResult
      }
    }

    this.setRegisterValue(context.registers, registerD, result)

    return new InstructionResult(-1)
  }
}

export class REM_U_64Instruction extends BaseInstruction {
  opcode: i32 = OPCODE_REM_U_64
  name: string = 'REM_U_64'
  
  execute(context: InstructionContext): InstructionResult {
    const registerD = this.getRegisterD(context.operands)
    const registerA = this.getRegisterA(context.operands)
    const registerB = this.getRegisterB(context.operands)
    const valueA = this.getRegisterValueAs64(context.registers, registerA)
    const valueB = this.getRegisterValueAs64(context.registers, registerB)

    // Gray Paper: when B = 0, result = A
    const result = valueB === i64(0) ? valueA : valueA % valueB

    this.setRegisterValue(context.registers, registerD, result)

    return new InstructionResult(-1)
  }
}

export class REM_S_64Instruction extends BaseInstruction {
  opcode: i32 = OPCODE_REM_S_64
  name: string = 'REM_S_64'
  
  execute(context: InstructionContext): InstructionResult {
    const registerD = this.getRegisterD(context.operands)
    const registerA = this.getRegisterA(context.operands)
    const registerB = this.getRegisterB(context.operands)
    const valueA = this.getRegisterValueAs64(context.registers, registerA)
    const valueB = this.getRegisterValueAs64(context.registers, registerB)

    // Convert to signed values using toSigned64 helper
    const signedA = this.toSigned64(valueA)
    const signedB = this.toSigned64(valueB)

    // Gray Paper: smod(a, b) = a when b = 0, or 0 when a = -2^63 and b = -1
    let result: i64
    if (signedA === -(i64(1) << i64(63)) && signedB === i64(-1)) {
      // Special overflow case: result = 0
      result = i64(0)
    } else if (signedB === i64(0)) {
      // Division by zero: result = a (signed modulo definition)
      result = valueA
    } else {
      // Normal case: sgn(a) * (|a| mod |b|)
      const absA = signedA < i64(0) ? -signedA : signedA
      const absB = signedB < i64(0) ? -signedB : signedB
      const sign = signedA < i64(0) ? i64(-1) : i64(1)
      const signedResult = sign * (absA % absB)
      // Convert signed result to unsigned using two's complement
      // Casting negative i64 to u64 automatically adds 2^64
      result = i64(u64(signedResult))
    }

    this.setRegisterValue(context.registers, registerD, result)

    return new InstructionResult(-1)
  }
}

File: packages/pvm-rust/src/instructions/arithmetic_64.rs
//! 64-bit register arithmetic (mirrors assembly/instructions/arithmetic-64.ts).
//! ADD_64, SUB_64, MUL_64, DIV_U_64, DIV_S_64, REM_U_64, REM_S_64.

use crate::config::{
    OPCODE_ADD_64, OPCODE_DIV_S_64, OPCODE_DIV_U_64, OPCODE_MUL_64, OPCODE_REM_S_64,
    OPCODE_REM_U_64, OPCODE_SUB_64,
};
use crate::instructions::base::{parse_three_registers, InstructionHandler};
use crate::types::{InstructionContext, InstructionResult};

fn get_register(registers: &[u64; 13], index: u8) -> u64 {
    registers.get(index as usize).copied().unwrap_or(0)
}

fn set_register(registers: &mut [u64; 13], index: u8, value: u64) {
    if (index as usize) < 13 {
        registers[index as usize] = value;
    }
}

fn to_signed64(x: u64) -> i64 {
    x as i64
}

// --- ADD_64 (200) ---
pub struct Add64Instruction;

impl Add64Instruction {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}

impl InstructionHandler for Add64Instruction {
    fn opcode(&self) -> i32 {
        i32::from(OPCODE_ADD_64)
    }
    fn name(&self) -> &'static str {
        "ADD_64"
    }
    fn execute(&self, context: &mut InstructionContext<'_>) -> InstructionResult {
        let parsed = parse_three_registers(context.operands);
        let value_a = get_register(context.registers, parsed.register_a);
        let value_b = get_register(context.registers, parsed.register_b);
        set_register(context.registers, parsed.register_d, value_a.wrapping_add(value_b));
        InstructionResult::new(InstructionResult::CONTINUE, 0)
    }
}

// --- SUB_64 (201) ---
pub struct Sub64Instruction;

impl Sub64Instruction {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}

impl InstructionHandler for Sub64Instruction {
    fn opcode(&self) -> i32 {
        i32::from(OPCODE_SUB_64)
    }
    fn name(&self) -> &'static str {
        "SUB_64"
    }
    fn execute(&self, context: &mut InstructionContext<'_>) -> InstructionResult {
        let parsed = parse_three_registers(context.operands);
        let value_a = get_register(context.registers, parsed.register_a);
        let value_b = get_register(context.registers, parsed.register_b);
        set_register(context.registers, parsed.register_d, value_a.wrapping_sub(value_b));
        InstructionResult::new(InstructionResult::CONTINUE, 0)
    }
}

// --- MUL_64 (202) ---
pub struct Mul64Instruction;

impl Mul64Instruction {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}

impl InstructionHandler for Mul64Instruction {
    fn opcode(&self) -> i32 {
        i32::from(OPCODE_MUL_64)
    }
    fn name(&self) -> &'static str {
        "MUL_64"
    }
    fn execute(&self, context: &mut InstructionContext<'_>) -> InstructionResult {
        let parsed = parse_three_registers(context.operands);
        let value_a = get_register(context.registers, parsed.register_a);
        let value_b = get_register(context.registers, parsed.register_b);
        set_register(context.registers, parsed.register_d, value_a.wrapping_mul(value_b));
        InstructionResult::new(InstructionResult::CONTINUE, 0)
    }
}

// --- DIV_U_64 (203) ---
pub struct DivU64Instruction;

impl DivU64Instruction {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}

impl InstructionHandler for DivU64Instruction {
    fn opcode(&self) -> i32 {
        i32::from(OPCODE_DIV_U_64)
    }
    fn name(&self) -> &'static str {
        "DIV_U_64"
    }
    fn execute(&self, context: &mut InstructionContext<'_>) -> InstructionResult {
        let parsed = parse_three_registers(context.operands);
        let value_a = get_register(context.registers, parsed.register_a);
        let value_b = get_register(context.registers, parsed.register_b);
        let result = if value_b == 0 {
            0xffff_ffff_ffff_ffff
        } else {
            value_a / value_b
        };
        set_register(context.registers, parsed.register_d, result);
        InstructionResult::new(InstructionResult::CONTINUE, 0)
    }
}

// --- DIV_S_64 (204) ---
pub struct DivS64Instruction;

impl DivS64Instruction {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}

impl InstructionHandler for DivS64Instruction {
    fn opcode(&self) -> i32 {
        i32::from(OPCODE_DIV_S_64)
    }
    fn name(&self) -> &'static str {
        "DIV_S_64"
    }
    fn execute(&self, context: &mut InstructionContext<'_>) -> InstructionResult {
        let parsed = parse_three_registers(context.operands);
        let value_a = get_register(context.registers, parsed.register_a);
        let value_b = get_register(context.registers, parsed.register_b);
        let signed_a = to_signed64(value_a);
        let signed_b = to_signed64(value_b);
        let result = if signed_b == 0 {
            0xffff_ffff_ffff_ffffu64
        } else if signed_a == i64::MIN && signed_b == -1 {
            value_a
        } else {
            let q = signed_a / signed_b;
            q as u64
        };
        set_register(context.registers, parsed.register_d, result);
        InstructionResult::new(InstructionResult::CONTINUE, 0)
    }
}

// --- REM_U_64 (205) ---
pub struct RemU64Instruction;

impl RemU64Instruction {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}

impl InstructionHandler for RemU64Instruction {
    fn opcode(&self) -> i32 {
        i32::from(OPCODE_REM_U_64)
    }
    fn name(&self) -> &'static str {
        "REM_U_64"
    }
    fn execute(&self, context: &mut InstructionContext<'_>) -> InstructionResult {
        let parsed = parse_three_registers(context.operands);
        let value_a = get_register(context.registers, parsed.register_a);
        let value_b = get_register(context.registers, parsed.register_b);
        let result = if value_b == 0 {
            value_a
        } else {
            value_a % value_b
        };
        set_register(context.registers, parsed.register_d, result);
        InstructionResult::new(InstructionResult::CONTINUE, 0)
    }
}

// --- REM_S_64 (206) ---
pub struct RemS64Instruction;

impl RemS64Instruction {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}

impl InstructionHandler for RemS64Instruction {
    fn opcode(&self) -> i32 {
        i32::from(OPCODE_REM_S_64)
    }
    fn name(&self) -> &'static str {
        "REM_S_64"
    }
    fn execute(&self, context: &mut InstructionContext<'_>) -> InstructionResult {
        let parsed = parse_three_registers(context.operands);
        let value_a = get_register(context.registers, parsed.register_a);
        let value_b = get_register(context.registers, parsed.register_b);
        let signed_a = to_signed64(value_a);
        let signed_b = to_signed64(value_b);
        let result = if signed_a == i64::MIN && signed_b == -1 {
            0u64
        } else if signed_b == 0 {
            value_a
        } else {
            let abs_a = signed_a.unsigned_abs();
            let abs_b = signed_b.unsigned_abs();
            let sign = if signed_a < 0 { -1i64 } else { 1 };
            let signed_rem = sign * (abs_a as i64 % abs_b as i64);
            signed_rem as u64
        };
        set_register(context.registers, parsed.register_d, result);
        InstructionResult::new(InstructionResult::CONTINUE, 0)
    }
}
